service/controller/resource/monitoring/remotewriteagentsecret/resource.go

```diff
@@ -0,0 +1,101 @@
+package remotewriteagentsecret
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/k8sclient/v7/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/prometheus-meta-operator/v2/pkg/password"
+	"github.com/giantswarm/prometheus-meta-operator/v2/service/controller/resource/generic"
+	"github.com/giantswarm/prometheus-meta-operator/v2/service/key"
+)
+
+const (
+	Name = "remotewriteagentsecret"
+)
+
+type Config struct {
+	K8sClient       k8sclient.Interface
+	Logger          micrologger.Logger
+	PasswordManager password.Manager
+}
+
+func New(config Config) (*generic.Resource, error) {
+	clientFunc := func(namespace string) generic.Interface {
+		c := config.K8sClient.K8sClient().CoreV1().Secrets(namespace)
+		return wrappedClient{client: c}
+	}
+
+	c := generic.Config{
+		ClientFunc:    clientFunc,
+		Logger:        config.Logger,
+		Name:          Name,
+		GetObjectMeta: getObjectMeta,
+		GetDesiredObject: func(ctx context.Context, v interface{}) (metav1.Object, error) {
+			return toSecret(ctx, v, config)
+		},
+		HasChangedFunc: hasChanged,
+	}
+	r, err := generic.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return r, nil
+}
+
+func getObjectMeta(ctx context.Context, v interface{}) (metav1.ObjectMeta, error) {
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return metav1.ObjectMeta{}, microerror.Mask(err)
+	}
+
+	return metav1.ObjectMeta{
+		Name:      key.RemoteWriteAgentSecretName,
+		Namespace: key.Namespace(cluster),
+	}, nil
+}
+
+func toSecret(ctx context.Context, v interface{}, config Config) (*corev1.Secret, error) {
+	objectMeta, err := getObjectMeta(ctx, v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	cluster, err := key.ToCluster(v)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	username := key.ClusterID(cluster)
+
+	password, err := config.PasswordManager.GeneratePassword(32)
+	if err != nil {
+		return nil, microerror.Mask(err)
```

Comment:
Can you put more context, on what's failing here?

Reply:
What do you mean by context? Can you give an example?

Reply:
If the controller fails in this stage, i don't know whats makes it fails.
If we add something like `error: password generation fails, err`:  
this gives me better view.

Reply:
Sure I can log something. My thinking is the error should be self-sufficient but I can add logs

Reply:
you can added it in the underlying function. if you want.

Reply:
It's added